Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,130 @@
+
+With only this change, Prometheus Agents can already be horizontally scaled without problems, but for Prometheus Servers it gets a little more complicated.
+
+### Graceful scale down of Prometheus Servers
+
+Prometheus Servers don't behave like the usual stateless applications. A few cases require Prometheus pods to still be available even after a scale down event is triggered:
+
+* Thanos proxying queries back to the sidecars to retrieve data that still wasn't uploaded to object storage.
+* Prometheus servers with long retention periods still hold data that users might want to query.
+
+If we simply delete Prometheus pods during scale-down, data loss will be a common problem, so a more complex scale-down strategy needs to be used here.
+
+We propose that Prometheus-Operator introduces a new field to the Prometheus Server CRDs:
+
+```yaml
+apiVersion: monitoring.coreos.com
+kind: Prometheus
+metadata:
+  name: example
+spec:
+  autoscaling:
+    scaleDownStrategy: force|preventDataLoss # Default: force
+```
+
+If scaleDownStrategy == force, the operator would simply delete the pod on the spot, not caring about the data stored in that particular shard. Although not the safest approach, it is exactly the current behavior nowadays and changing it might surprise a lot of users.
+
+If scaleDownStrategy == preventDataLoss then Prometheus-Operator would first mark the pod with an annotation:
+
+```yaml
+prometheus.operator.io/safe-deletion-timestamp: X
+```
+
+Where X is the timestamp of the current retention period.
+
+The configuration of all shard would be re-arranged, to make sure the Prometheus that was scaled down do not scrapes any metrics anymore, but it would still be available for queries.
+
+When the current time finally surpasses the one in the annotation, the Operator will finally trigger its deletion.
+
+# Scale down Alternatives
+
+### Dump & Backfill
+
+During scale-down, Prometheus-Operator could read TSDB Blocks from the Prometheus instance being deleted and backfill it into another instance.
+
+***Advantages:***
+* Prometheus can be shutdown immediately without data loss.
+
+***Disadvantages:***
+* Loading TSDB Blocks into memory is expensive, requiring Prometheus-Operator to run with big Memory/CPU requests.
+
+### Snapshot & Upload on shutdown
```

Comment:
I think that it might be a viable alternative for folks using Prometheus agent + Thanos as their main metrics storage.

Reply:
Prometheus Server with Thanos Sidecar as well no?

Reply:
You mean Thanos sidecar with object upload?

Reply:
Yup

Reply:
I'm not familiar enough but you might still want to keep the pods around for the retention period in case you have local alerting?